### DISCUSSION TITLE
ProjectionAux in parallel with lower dimensional elements

### DISCUSSION THREAD

**Discussion by mcacace** [idaholab/moose#28599] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Dear moose team, 
I'm trying to convert an elemental auxvariable (storing a material property) onto a nodal auxvariable via ProjectionAux for cases where I have multiple lower dimensional elements in my simulation (3D, 2D and 1D). The error only occurs if I run on multiple nodes. I tried to debug in parallel, but after setting breakpoints, the simulations stops without giving me the chance to get the backtrace and the only message I have reads 

>  libMesh terminating:
>
> [1] ../src/systems/system.C, line 2440, compiled May 15 2024 at 11:55:37

I tried to have a look at the system class in libmesh, and it might point to a problem inside `Number System::point_value` (though I am relying on a conda installation locally and cannot tell if the files are the same).  
Any help? Thanks for any feedback
On a side note: I created a very simple file only relying on the framework objects, and I can share if needed.

**Comment by mcacace** [comment]
Forgot to add that without debugger the error is segfault (not sure to which libmesh_assert calls it does refer to)

**Reply by mcacace** [reply]
Manage to have the bt (attached) and indeed it points to the `Number System::point_value` call. Sorry for multiple posting
[bt.log](https://github.com/user-attachments/files/16962831/bt.log)

**Reply by GiudGiud** [reply]
The ProjectionAux has not been fitted to handle lower D elements. We can take a look if you have a minimal example, to see if it can be expanded

**Reply by mcacace** [reply]
@GiudGiud thanks! My simple test case sums up to around 4k dofs. If you think is doable (in my case it debug over seconds and the segfault only occurs with >=3 processors) I can share it with you (mesh+input file). What would be your preferred options? Thanks for looking into it

**Reply by GiudGiud** [reply]
4k is a little much, we ll want something that should be useable as a test after

**Reply by mcacace** [reply]
Ok, tomorrow will try to set up a simple model.

**Reply by mcacace** [reply]
@GiudGiud I create a minimal example (17 dofs).
My suspect is that the error occurs once you have multiple families of LDs (2D and 1D in my test case).
Please, let me know how do you prefer for me to share the input and mesh file (sorry I could not find any MeshGenerator nor gmsh functions to have all LDs inside). 
Thanks again for helping

**Reply by GiudGiud** [reply]
I think lower-D is immediately a problem, the other variable would not be defined on the lower-D element.

you could put the two files on a public github repo of yours and paste a link here for example

**Reply by mcacace** [reply]
Sorry for being late. Here is the github repo [(projectionAux)](https://github.com/mcacace/ProjectionAux).
Thanks again for helping

**Reply by GiudGiud** [reply]
I m not hitting a seg fault on this example. 
Are there lower D elements in this example? I see a nodal and an elemental variable but they are not block restricted so they would be defined everywhere

**Reply by mcacace** [reply]
There should be a 2D and 1D LD elements. Is it possible that you are running on a single processor? The segfault usually occurs while running over multiple processors. And yes, the auxvariables are defined for all blocks, 3D, 2D and 1D.

**Reply by mcacace** [reply]
Just checked and indeed by running on a single processor you won't hit the segfault.

**Reply by GiudGiud** [reply]
I can reproduce the error. 
What behavior are you trying to achieve? Projecting inside the lower D? (so lower D variable values to lower D variable values?)
or projecting from higher D to lower D ? or lower D to higher D?

**Reply by mcacace** [reply]
The idea is a simple projection from elemental to nodal values, though typically more than one order of elements will share a node. In principle, I do not require higher to lower and the other way around, this I do over the same family of elements via a custom AuxKernel. 
I also tried to avoid LD elements in the projection by block-wise both auxvariables/kernels to the 3D variables but still received the seg fault, it seems that the instantiation of the libmesh System object does work over the entire mesh, though I might be wrong here.

**Reply by mcacace** [reply]
by block-wise both auxvariables/kernels to the 3D **elements only**
sorry

**Reply by GiudGiud** [reply]
ok I have a patch here that uses only higher dimensional elements
https://github.com/idaholab/moose/pull/28631

I am curious what people's needs would be for lower dimensional elements. i am banning it for now as I dont consider it's supported

**Reply by mcacace** [reply]
@GiudGiud I see ... thanks for the feedback.

**Reply by mcacace** [reply]
@GiudGiud should the elemOnNodeVariableIsDefinedOn be rather 

`for (const auto & elem_id : _mesh.nodeToElemMap().find(_current_node->id())->second)`
  ` {`
    **` const auto block_id = _mesh.elemPtr(elem_id)->subdomain_id();`**
    ` if (_source_variable.hasBlocks(block_id) &&`
       `  !_mesh.isLowerD(block_id) && hasBlocks(block_id))`
      ` return _mesh.elemPtr(elem_id);`
  ` }`

block_id is otherwise not defined here

**Reply by GiudGiud** [reply]
probably. I am amedning this a little right now

**Reply by GiudGiud** [reply, marked as answer]
ok done

Note that the new boolean I added to exclude additonal blocks should be used here

otherwise when projecting you will use the nodal value from the high D block and from the lower D block (in the summation to compute the local average value)

**Reply by mcacace** [reply]
Excellent! thanks @GiudGiud I will close the post then.

**Reply by GiudGiud** [reply]
I m not getting consistent results in parallel on that example mesh.
You'll need to be watchful on your mesh for now. It will run but results might not be the true projection, it might be excluding values it should not

**Reply by mcacace** [reply]
Thanks! I'm testing and it does work OK as far as I can tell based on my own (different) example. I will report back if encountering any specific issue.

**Reply by GiudGiud** [reply]
turns out it was just a numbering issue. so should be good (I ll message again here if not)

**Reply by GiudGiud** [reply]
So it turns out it started diffing again at higher number of cores (starting at 5, then randomly up to 16).
I removed the 1D elements from cube.e then got consistent results. 

Is this 1D, 2D and 3D elements at the same time (and all attached) situation what you need for your simulations?

**Reply by mcacace** [reply]
Thanks for checking. It was indeed my feeling that the problem starts if either more than one order of LD elements are there, or if the dimensions between the elements is >1 (1D vs 3D). I'm currently out of office (back 30th of September) and couldn't check the latter. And yes, we routinely make use of all 3 orders of elements in our runs.
